### ISSUE TITLE
New design ideas

### ISSUE THREAD

**Issue by franleplant** [franleplant/reform#14]
I've been starting to work on React-Native and I've seen that the notion of `form` does not exist, so there will be some fixes and adaptations to be made to enable `reform` to work on mobile.

While I was thinking of it I thought of and idea to improve and simplify the interface of `reform`. The core idea will be to have a _by control_ function to use.

```javascript

<input
  required
  value={this.state.fields.myControl}
  {...reform(myOnchangeHandler)}
/>
```

This will desugar to something like this
```javascript
<input
  required
  value={this.state.fields.myControl}
  onChange={reform(myOnChangeHandler)}
  ... (potentially other useful attrs)
/>
```

The core idea is that we can access the control props in the onChange handler (we need to test the limits of this techniques, because it might be too limiting).

Internally, `reform` function will be just a proxy `changeHandler`, something like this

```javascript

function reform(onChangeHandler, options) {
  return (e, ...args) => {
    const nativeValidationRules = parseNativeValidationRules(e);
    const customValidationRules = getCustomValidationRules(options);
    const rules = safeMerge(nativeValidationRules, customValidationRules);
    const value = options.getValue(e);

    const errorMap = calcErrors(rules, value);

    onChangeHandler.call(null, errorMap, e, ...args);
  }
}
```

There are some questions here:
- How can we work with custom components that do not call with a proper event?
- What are the limititations of event parsing
- et al I think

I will update this as I get the chance to think more about it.

**Comment by franleplant** [comment]
We could go even more apart from HTML5 and use something like this
```javascript
reform(onChangeHandlers, rules, getValue, ...)
```

Good:
- very simple interface, I think I can reduce the code by a lot
- reform becomes stateless, effectively being just a onChange factory
- it composes well with basically anything because you don't need to use it in all form controls


BAD
- we are diverging from the HTML5 api (that might not be a bad thing but...)
- potential limitations?
- big api change

**Comment by franleplant** [comment]
I've been working on Form Validation on a React native project and I've implemented a Ad Hoc solution with very few lines of code that worked really well. I think that briging that experience to Reform will be very good. This are the things implemented in it:

- validator map (with the usual value => boolean signature), with a few common validators such as required and email
- mapHasError and mapMapHasError useful pure functions that take an ErrorMap or a Map of ErrorMaps and tell you weather some of it's fields are true which will indicate that the Form Field which that ErrorMap belongs to is invalid.
- validate function. This is the core of it all. It receives a value and rules and return an ErrorMap.

I've also used some kind of _mixin_ function that reduce the boilerplate but those could even be optional or made ad hoc by the user.

One particularly important thing I've noticed is that since React Native only has a `TextInput` and few other controls, the interface is pretty minimal, compared to the multiple permutations of `input` and `type` that HTML5 exposes. So maybe we could approach this with a mix of both, maintaining simplicity but also provinding similar validation directives as HTML5.

This could be done by smart naming the validator rules.

For example:

- `minNumber: 10` -> will be equal to `<input type=number min=10/>`
- `minDate: 10/10/2016` -> will be equal to `<input type=date min=10/10/2016/>`

So everybody win.

The interface is arguable more verbose than HTML5 but it has the great advantage of making everything simpler and everything super customizable to the user's need.

So, for example, the validate function could be use `onBlur`, `onChange`, or on any event user wants, additionally, we can remove the necesity of specifying `getValue` by simply doing this:

```javascript
onChange={(e) => {
  const errorMap = validate(e.target.value, rules)
  // store errorMap in state
}}

```

If user abstracts `validate` to also do all the error storing and whatever other stuff he might need then the interface is easier



```javascript
onChange={e => this.myValidate(e.target.value, rules)}
```

Note that rules accept functions, in the form of ad hoc custom validators :)


As you see, it's pretty easy.


So the gist of the plan to implement this is to:

1- Create a minimal typescript/webpack build system
2- Create (probably in a standalone package) the validators that match HTML5 validators
3- Create the rest of the functions: the validator/rule evaluator and else
4- Test with jest
5- Find out how this work with radio buttons since they are super complicated. We could let the user handle this in her own way but I'd like a comprehensive solution.
